File: sdk/java/touca/src/main/java/io/touca/core/Transport.java
// Copyright 2023 Touca, Inc. Subject to Apache-2.0 License.

package io.touca.core;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;

/**
 * Contains logic for communicating with the Touca server.
 */
public final class Transport {
  private String apiKey;
  private String apiUrl;

  /**
   * Response of an HTTP request including status code and response body.
   */
  public static final class Response {
    public int code;
    public String content;

    Response(final int code, final String content) {
      this.code = code;
      this.content = content;
    }
  }

  private String readResponse(final InputStream inputStream)
      throws IOException {
    final StringBuilder builder = new StringBuilder();
    try (BufferedReader reader = new BufferedReader(
        new InputStreamReader(inputStream, StandardCharsets.UTF_8))) {
      for (String line; (line = reader.readLine()) != null;) {
        builder.append(line);
      }
    }
    return builder.toString();
  }

  private HttpURLConnection makeConnection(final String path)
      throws IOException {
    final URL url = new URL(this.apiUrl + path);
    final HttpURLConnection con = (HttpURLConnection) url.openConnection();
    con.setRequestProperty("Accept", "application/json");
    con.setRequestProperty("Accept-Charset", "utf-8");
    con.setRequestProperty("User-Agent", "touca-client-java/1.7.1");
    con.setRequestProperty("X-Touca-API-Key", this.apiKey);
    return con;
  }

  /**
   * Attempt to authenticate with the Touca server.
   */
  public void configure(final String apiUrl, final String apiKey) {
    if (this.apiUrl == apiUrl && this.apiKey == apiKey) {
      return;
    }
    this.apiUrl = apiUrl;
    this.apiKey = apiKey;
    final Response response = postRequest("/client/verify", "application/json",
        new byte[0]);
    if (response.code == HttpURLConnection.HTTP_UNAUTHORIZED) {
      throw new ToucaException("authentication failed: API key invalid");
    } else if (response.code != HttpURLConnection.HTTP_NO_CONTENT) {
      throw new ToucaException("authentication failed: invalid response");
    }
  }

  /**
   * Submits an HTTP GET request to a specific API endpoint.
   *
   * @param path API endpoint to submit HTTP request to
   * @return response received from the server
   */
  public Response getRequest(final String path) {
    try {
      final HttpURLConnection con = makeConnection(path);
      con.setRequestMethod("GET");
      return new Response(con.getResponseCode(),
          readResponse(con.getResponseCode() < 400 ? con.getInputStream() : con.getErrorStream()));
    } catch (final IOException ex) {
      throw new ToucaException(ex.getMessage());
    }
  }

  /**
   * Submits an HTTP POST request to a specific API endpoint.
   *
   * @param path        API endpoint to submit HTTP request to
   * @param contentType either "application/json" or "application/octet-stream"
   * @param content     content in bytes to submit to the endpoint
   * @param headers     headers to add to the HTTP request
   * @return response received from the server
   */
  public Response postRequest(final String path, final String contentType, final byte[] content,
      final Map<String, String> headers) {
    try {
      final HttpURLConnection con = makeConnection(path);
      con.setDoOutput(true);
      con.setRequestMethod("POST");
      con.setRequestProperty("Content-Type", contentType);
      headers.forEach((k, v) -> con.setRequestProperty(k, v));
      con.connect();
      if (content.length != 0) {
        try (OutputStream stream = con.getOutputStream()) {
          stream.write(content, 0, content.length);
        }
      }
      return new Response(con.getResponseCode(),
          readResponse(con.getResponseCode() < 400 ? con.getInputStream() : con.getErrorStream()));
    } catch (final IOException ex) {
      throw new ToucaException(ex.getMessage());
    }
  }

  public Response postRequest(final String path, final String contentType, final byte[] content) {
    return postRequest(path, contentType, content, new HashMap<>());
  }
}

File: sdk/cpp/src/transport.cpp
// Copyright 2023 Touca, Inc. Subject to Apache-2.0 License.

#include "touca/core/transport.hpp"

#include <regex>
#include <sstream>

#include "httplib.h"
#include "rapidjson/document.h"
#include "touca/core/filesystem.hpp"

namespace touca {

void DefaultTransport::configure(const std::string& api_key,
                                 const std::string& api_url) {
  const auto& userAgent =
      touca::detail::format("touca-client-cpp/{}.{}.{}", TOUCA_VERSION_MAJOR,
                            TOUCA_VERSION_MINOR, TOUCA_VERSION_PATCH);
  _api_url = ApiUrl(api_url);
  _cli = touca::detail::make_unique<httplib::Client>(_api_url.root.c_str());
  _cli->set_default_headers({{"Accept-Charset", "utf-8"},
                             {"Accept", "application/json"},
                             {"User-Agent", userAgent},
                             {"X-Touca-API-Key", api_key}});
#ifdef CPPHTTPLIB_OPENSSL_SUPPORT
  _cli->enable_server_certificate_verification(false);
#endif
  const auto response = post("/client/verify");
  if (response.status == -1) {
    throw touca::detail::runtime_error(
        touca::detail::format("failed to reach server at {}", api_url));
  }
  if (response.status == 401) {
    throw touca::detail::runtime_error(
        "Authentication failed: API Key Invalid.");
  }
  if (response.status != 204) {
    throw touca::detail::runtime_error(touca::detail::format(
        "Authentication failed: Invalid Response: {}", response.status));
  }
}

Response DefaultTransport::get(const std::string& route) const {
  const auto& result = _cli->Get(_api_url.route(route).c_str());
  if (!result) {
    return {-1, touca::detail::format("failed to submit HTTP GET request to {}",
                                      route)};
  }
  return {result->status, result->body};
}

Response DefaultTransport::patch(const std::string& route,
                                 const std::string& body) const {
  const auto& result =
      _cli->Patch(_api_url.route(route).c_str(), body, "application/json");
  if (!result) {
    return {-1, touca::detail::format(
                    "failed to submit HTTP PATCH request to {}", route)};
  }
  return {result->status, result->body};
}

Response DefaultTransport::post(const std::string& route,
                                const std::string& body) const {
  const auto& result =
      _cli->Post(_api_url.route(route).c_str(), body, "application/json");
  if (!result) {
    return {-1, touca::detail::format(
                    "failed to submit HTTP POST request to {}", route)};
  }
  return {result->status, result->body};
}

Response DefaultTransport::binary(const std::string& route,
                                  const std::string& content,
                                  const Transport::Headers& headers) const {
  httplib::Headers extra;
  for (const auto& header : headers) {
    extra.emplace(header);
  }
  const auto& result = _cli->Post(_api_url.route(route).c_str(), extra, content,
                                  "application/octet-stream");
  if (!result) {
    return {-1, touca::detail::format(
                    "failed to submit HTTP POST request to {}", route)};
  }
  return {result->status, result->body};
}

DefaultTransport::DefaultTransport() : _api_url(""), _cli(nullptr) {}

DefaultTransport::~DefaultTransport() {}

ApiUrl::ApiUrl(const std::string& api_url) {
  if (api_url.empty()) {
    return;
  }
  const static std::regex pattern(
      R"(^(?:([a-z]+)://)?([^:/?#]+)(?::(\d+))?/?(.*)?$)");
  std::cmatch result;
  if (!std::regex_match(api_url.c_str(), result, pattern)) {
    throw touca::detail::runtime_error(
        touca::detail::format("API URL has invalid format: {}", api_url));
  }

  std::vector<std::string> elements = {result[1], result[2], result[3]};
  const std::string path = result[4];
  if (!path.empty()) {
    const auto index = path.find_last_of('@');
    prefix = path.substr(0, index);
    if (!prefix.empty() && prefix.back() == '/') {
      prefix.pop_back();
    }
    if (index != std::string::npos) {
      extra = path.substr(index + 1);
    }
  }

  root = elements[1];
  if (!elements[0].empty()) {
    root.insert(0, touca::detail::format("{}://", elements[0]));
  }
  if (!elements[2].empty()) {
    root.append(touca::detail::format(":{}", elements[2]));
  }
}

std::string ApiUrl::route(const std::string& path) const {
  return prefix.empty() ? path : "/" + prefix + path;
}

}  // namespace touca
